run command new commit (#271)
Azure/arc_jumpstart_docs: docs/azure_arc_jumpstart/azure_arc_servers/day2/arc_run_command/_index.md

---
type: docs

> title: "Azure ARC RUN command"

Reviewer: "Arc" instead of "ARC"

Author: done

> linkTitle: "Azure ARC RUN command"

Reviewer: "Arc" instead of "ARC"

Author: done

weight: 20
description: >
---

> ## Run Powershell and Shell scripts on Azure Arc-enabled servers using the Run command

Reviewer: "PowerShell" instead of "Powershell"

Author: done

> The Run command feature uses the Connected Machine agent to remotely run PowerShell scripts within an Azure Arc-connected Windows machine and Shell scripts within an Azure Arc-connected Linux machine. 

Reviewer 2: add a link to the feature in MSLearn

> The Run command feature uses the Connected Machine agent to remotely run PowerShell scripts within an Azure Arc-connected Windows machine and Shell scripts within an Azure Arc-connected Linux machine. 

Reviewer 2: Add link to feature page in MSLearn

Author: added

> The Run command feature uses the Connected Machine agent to remotely run PowerShell scripts within an Azure Arc-connected Windows machine and Shell scripts within an Azure Arc-connected Linux machine. 
> 
> This capability is useful in all scenarios where you want to run a script within an Arc-connected machine. It allows you to troubleshoot and remediate a machine that doesn't have the RDP or SSH port open because of improper network or administrative user configuration.
> 
> The following Jumpstart scenario will guide you on how to use the Run command on your Arc-connected machine. Use the **Azure Cloud Shell** or **Visual Studio Code** command terminal to follow the instructions in this scenario.

Reviewer 2: ```suggestion
The Run command feature uses the Connected Machine agent to remotely run PowerShell scripts within an Azure Arc-connected Windows machine and Shell scripts within an Azure Arc-connected Linux machine. This capability is useful in all scenarios where you want to run a script within an Arc-connected machine. It allows you to troubleshoot and remediate a machine that doesn't have the RDP or SSH port open because of improper network or administrative user configuration.

The following Jumpstart scenario will guide you on how to use the Run command on your Arc-connected machine. Use the **Azure Cloud Shell** or **Visual Studio Code** command terminal to follow the instructions in this scenario.
```

Author: done

- **[GCP Ubuntu instance](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/gcp/gcp_terraform_ubuntu/)**
- **[GCP Windows instance](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/gcp/gcp_terraform_windows/)**
- **[AWS Ubuntu EC2 instance](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/aws/aws_terraform_ubuntu/)**
- **[AWS Amazon Linux 2 EC2 instance](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/aws/aws_terraform_al2/)**
- **[Azure Ubuntu VM](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/azure/azure_arm_template_linux/)**
- **[Azure Windows VM](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/azure/azure_arm_template_win/)**
- **[VMware vSphere Ubuntu VM](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/vmware/vmware_terraform_ubuntu/)**
- **[VMware vSphere Windows Server VM](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/vmware/vmware_terraform_winsrv/)**
- **[Vagrant Ubuntu box](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/vagrant/local_vagrant_ubuntu/)**

> - **[Vagrant Windows box](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_servers/vagrant/local_vagrant_windows/)**

Reviewer 2: These are old addresses. change io to dot com and validate

Author: done and verified

## Prerequisites

- As mentioned, this scenario starts at the point where you already deployed and connected VMs or servers to Azure Arc. In the screenshots below, you can see a Windows and a Linux server that have been connected with Azure Arc and are visible as resources in Azure.

    ![Screenshot Azure Arc-enabled servers on resource group](./01.png)

    ![Screenshot Linux Azure Arc-enabled server connected status](./02.png)

    ![Screenshot Windows Azure Arc-enabled server connected status](./03.png)

> - [Install or update Azure CLI](https://docs.microsoft.com/cli/azure/install-azure-cli?view=azure-cli-latest) on your client machine. Azure CLI should be running version 2.49.0 or later. Use ```az --version``` to check your current installed version.

Reviewer 2: ```suggestion
- [Install or update Azure CLI](https://learn.microsoft.com/cli/azure/install-azure-cli?view=azure-cli-latest) on your client machine. Azure CLI should be running version 2.49.0 or later. Use ```az --version``` to check your current installed version.
```

Author: done

```
- If the version of the Azure cli extension is older than "0.5.1" then use the following commands to upgrade it 
```shell
    az extension remove --name connectedmachine
    az extension add -- name connectedmachine
```
- Check the version of the Connected Machine agent running on your Arc-connected machine using the following command

```shell
    az connectedmachine show --name <Machine Name> --resource-group <Resource Group Name> --query "agentVersion"
```
- If the Connected Machine agent version is lower than 1.39 then upgrade the extension using [this guidance](https://learn.microsoft.com/en-us/azure/azure-arc/servers/manage-agent).

- To use PowerShell you need to check if the installed version of module _Az.ConnectedMachine_ is 0.6.0 or higher. Use the following PowerShell command to check the installed version

```powershell
Get-Module -ListAvailable -Name Az.ConnectedMachine
```
Use the following command to install the latest version of the PowerShell module

```powershell
Install-Module -Name Az.ConnectedMachine -Force 
```

## Use the Run command to execute a simple PowerShell command within an Arc-connected Windows machine

> - Run the following Azure cli command after adding the appropriate resource group, name of the Arc-connected machine, a name identifying the command and the location of you Arc-connected machine

Reviewer 2: ```suggestion
- Run the following Azure CLI command after adding the appropriate resource group, name of the Arc-connected machine, a name identifying the command, and the location of your Arc-connected machine
```

Author: changed all cli to CLI

```shell
    az connectedmachine run-command create --resource-group <Resource Group Name> --machine-name <Machine Name> --run-command-name <Identifying Name of command> --script "Write-Host 'Hello World'" --location <Location>
```

> - It takes few minutes to run the command and return the results. If the execution is successful then you would see the following within the longer returned Json string:

Reviewer 2: ```suggestion
- It takes a few minutes to run the command and return the results. If the execution is successful then you will see the following within the longer returned Json string:
```

Author: done

    
```shell
    "executionState": "Succeeded",
    "exitCode": 0,
    "output": "Hello World",
```

> If you prefer to use PowerShell then use the following command

Reviewer 2: ```suggestion
If you prefer to use PowerShell then use the following command:
```

Author: done

```powershell
New-AzConnectedMachineRunCommand -ResourceGroupName "<Resource Group Name>" -Location "<Location>" -SourceScript "Write-Host 'Hello World'" -RunCommandName "<Identifying Name of command>" -MachineName "<Machine Name>"
```

> The successful output of the PowerShell command will show the following output

Reviewer 2: ```suggestion
The successful output of the PowerShell command will show the following output:
```

Author: done

> ![Screenshot success helloworld PowerShell](07.png)

Reviewer 2: 07 comes before 06.png. Align file names according to it's position in the doc.

Author: Fixed the screenshots order in the document

## Use the Run command to execute a simple Shell command within an Arc-connected Linux machine

> - On your client machine run the following Azure cli command after adding the appropriate parameters

Reviewer 2: ```suggestion
- On your client machine run the following Azure CLI command after adding the appropriate parameters
```

Author: cli to CLI changed in the doc

```shell
    az connectedmachine run-command create --resource-group <Resource Group Name> --machine-name <Machine Name> --run-command-name <Identifying Name of command> --script "ifconfig" --location <Location>
```

> Or in PowerShell

Reviewer 2: ```suggestion
Or in PowerShell:
```

Author: done

```powershell
New-AzConnectedMachineRunCommand -ResourceGroupName "<Resource Group Name>" -Location "<Location>" -SourceScript "ifconfig" -RunCommandName "<Identifying Name of command>" -MachineName "<Machine Name>"
```

> - If the execution is successful then you should have an output which includes the result of the ifconfig command as a string which includes "\n" to indicate line breaks in the original output.

Reviewer 2: ```suggestion
- If the execution is successful then you should have an output which includes the result of the _ifconfig_ command as a string which includes "\n" to indicate line breaks in the original output.
```

Author: done, re-edited

    ![Screenshot](./04.png)

## Examine the available options for the Run command

- Run the following command which returns the available options of the Run command such as "create", "delete", "list", "show", "update" and "wait".

```shell
    az connectedmachine run-command --help
```

> - Investigate the different ways you can run the "create" option by running the following command on your client machine

Reviewer 2: ```suggestion
- Investigate the different ways you can run the "create" option by running the following command on your client machine.
```

Reviewer 2: Please go over the readme and use dot at the end of sentences

Author: done, sometimes using : instead of .

```shell
    az connectedmachine run-command create --help
```

## Direct the output of a Run command to Azure storage blob

- Create a storage account (if you do not have one) using the following command after filling out the required parameters

```shell
    az storage account create --name <Storage account name> --resource-group <Resource Group Name> --location <Location> --sku Standard_LRS --kind storageV2 --allow-blob-public-access false
```
- Create a storage container to which you will direct the output of the run command

```shell
    az storage container create --name <container name> --account-name <Storage account name> --auth-mode login
```

> - Create a blob SAS URI with the following permissions: Read, Write, Create, delete and append. We will need an end date for the validity of the SAS token, 24 hours from the current date in the following example. Also, to be able to use the SAS URI in our run command we will need to remove any double quotes from the beginning and the end.

Author: done

```shell
    end=`date -u -d "24 hours" '+%Y-%m-%dT%H:%MZ'`
    sasuri=$(az storage blob generate-sas --account-name <storage account name> --container-name <storage container name> --name <name of blob for command output destination - it will be created if it doesn't exist> --permissions acdrw --expiry $end --full-uri | tr -d '"')

> ``````

Reviewer 2: ```suggestion
```
```

> ``````

Reviewer 2: ```suggestion
```
```

> ``````

Reviewer 2: only 3

Author: done

```powershell
    $end=`date -u -d "24 hours" '+%Y-%m-%dT%H:%MZ'`
    $sasuri=$(az storage blob generate-sas --account-name <storage account name> --container-name <storage container name> --name <name of blob for command output destination - it will be created if it doesn't exist> --permissions acdrw --expiry $end --full-uri | tr -d '"')
```

> - Execute the following run command which runs a PowerShell script within the Arc-enabled Windows machine. The run command directs the output to the append blob.

Reviewer 2: ```suggestion
- Execute the following _run_ command which runs a PowerShell script within the Arc-enabled Windows machine. The run command directs the output to the append blob.
```

Author: done
